src/details/ArborX_DetailsDistributedTreeNearest.hpp

```diff
@@ -108,16 +107,43 @@ struct AccessTraits<
 namespace Details
 {
 
-struct PairIndexDistance
+template <typename Value>
+struct PairValueDistance
 {
-  int index;
+  Value value;
   float distance;
 };
 
 template <typename Tree>
 struct CallbackWithDistance
 {
   Tree _tree;
+
+  template <typename ExecutionSpace>
+  CallbackWithDistance(ExecutionSpace const &, Tree const &tree)
+      : _tree(tree)
+  {}
+
+  template <typename Query, typename Value, typename Output>
+  KOKKOS_FUNCTION void operator()(Query const &query, Value const &value,
+                                  Output const &out) const
+  {
+    // FIXME no access to indexable getter
+    out({value, distance(getGeometry(query), _tree._indexable_getter(value))});
+  }
+};
+
+template <typename MemorySpace>
+struct CallbackWithDistance<BoundingVolumeHierarchy<
+    MemorySpace, Details::LegacyDefaultTemplateValue,
+    Details::DefaultIndexableGetter, ExperimentalHyperGeometry::Box<3, float>>>
+{
```

Comment:
The really bad. Can't get rid of the old legacy thingie just yet, as we cannot get a bounding volume by index.

Reply:
Why do we need both the primary template and that specialization?

Reply:
The primary template is based on a general tree, which really means the APIv2 tree. We have access to indexable and values, so don't need to store reverse permutation.

The specialization is for the APIv1 (specified through the default template parameters, including `LegacyDefaultTemplateValue`). We don't have access to indexable there. Even if we did, the problem is that this callback is wrapped into `LegacyCallbackWrapper` inside APIv1, so we don't have access to the full `PairValueIndex` value and can't call indexable on it. So, we have to do the other things.

This is pretty much unavoidable, as we *have* to use APIv1 `BottomTree` for APIv1 `DistributedTree` to be able to work with APIv1 post-callback interface in the distributed setting.